---
title: "infercnv_run"
author: "Brian Haas"
date: "9/5/2018"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(infercnv)

source("~/GITHUB/bhaas_forks/inferCNV/scripts/explore_steps_by_gene.simple.R")


```

## Create the InferCNV Object

```{r}
infercnv_obj = CreateInfercnvObject(
  raw_counts_matrix="oligodendroglioma_expression_downsampled.counts.matrix",
  annotations_file="oligodendroglioma_annotations_downsampled.txt",
  delim="\t",
  gene_order_file="gencode_downsampled.txt",
  ref_group_names=c("Microglia/Macrophage","Oligodendrocytes (non-malignant)"))

```

## Normalize each cell's counts for sequencing depth

```{r}
infercnv_obj <- infercnv:::normalize_counts_by_seq_depth(infercnv_obj)
```



```{r}
# filter out low expressed genes
cutoff=2
infercnv_obj <- infercnv:::require_above_min_mean_expr_cutoff(infercnv_obj, cutoff)
```


```{r}
# filter out bad cells
min_cells_per_gene=3
infercnv_obj <- infercnv:::require_above_min_cells_ref(infercnv_obj, min_cells_per_gene=min_cells_per_gene)
```


```{r}
## for safe keeping
infercnv_orig_filtered = infercnv_obj
plot_mean_chr_expr_lineplot(infercnv_obj)
save('infercnv_obj', file = 'infercnv_obj.orig_filtered')
```


## perform anscombe normalization
```{r}
infercnv_obj <- infercnv:::anscombe_transform(infercnv_obj)
save('infercnv_obj', file='infercnv_obj.anscombe')
plot_mean_chr_expr_lineplot(infercnv_obj)

```



## log transform the normalized counts:

```{r}
infercnv_obj <- infercnv:::log2xplus1(infercnv_obj)
save('infercnv_obj', file='infercnv_obj.log_transformed')
```




## perform smoothing across chromosomes

```{r}
infercnv_obj = infercnv:::smooth_by_chromosome(infercnv_obj, window_length=101, smooth_ends=TRUE)
save('infercnv_obj', file='infercnv_obj.smooth_by_chr')
plot_mean_chr_expr_lineplot(infercnv_obj)
```



```{r}
# re-center each cell
infercnv_obj <- infercnv:::center_cell_expr_across_chromosome(infercnv_obj, method = "median")
save('infercnv_obj', file='infercnv_obj.cells_recentered')
plot_mean_chr_expr_lineplot(infercnv_obj)
```


```{r}
infercnv:::plot_cnv(infercnv_obj, output_filename='infercnv.chr_smoothed', x.range="auto", title = "chr smoothed")
```
```{r}
knitr::include_graphics("infercnv.chr_smoothed.png")
```



# subtract the reference values from observations, NOW HAVE LOG (FOLD CHANGE) VALUES

```{r}
infercnv_obj <- infercnv:::subtract_ref_expr_from_obs(infercnv_obj)
save('infercnv_obj', file='infercnv_obj.ref_subtracted')
plot_mean_chr_expr_lineplot(infercnv_obj, sep_obs_types = TRUE)
```

```{r}
infercnv:::plot_cnv(infercnv_obj, output_filename='infercnv.ref_subtracted', x.range="auto", title="ref subtracted")
```

```{r}
knitr::include_graphics("infercnv.ref_subtracted.png")
```




## invert log values
```{r}

infercnv_obj@processed.data <- 2^infercnv_obj@processed.data
save('infercnv_obj', file='infercnv_obj.inverted_log')
plot_mean_chr_expr_lineplot(infercnv_obj, sep_obs_types = TRUE)
infercnv:::plot_cnv(infercnv_obj, output_filename='infercnv.inverted', color_safe_pal = FALSE, x.range="auto", x.center=1, title = "inverted log FC to FC")


```


```{r}
knitr::include_graphics("infercnv.inverted.png")
```

```{r}
# remove noise
infercnv_obj <- infercnv:::clear_noise_via_ref_mean_sd(infercnv_obj, sd_amplifier = 1.5)
save('infercnv_obj', file='infercnv_obj.denoised')
plot_mean_chr_expr_lineplot(infercnv_obj, sep_obs_types = TRUE)
infercnv:::plot_cnv(infercnv_obj, output_filename='infercnv.denoised', x.range="auto", x.center=1, title="denoised")
```

```{r}
knitr::include_graphics("infercnv.denoised.png")
```

## Remove outlier data points

```{r}
infercnv_obj = infercnv:::remove_outliers_norm(infercnv_obj)
save('infercnv_obj', file="infercnv_obj.outliers_removed")
infercnv:::plot_cnv(infercnv_obj, output_filename='infercnv.outliers_removed', color_safe_pal = FALSE, x.range="auto", x.center=1, title = "outliers removed")
```


```{r}
knitr::include_graphics("infercnv.outliers_removed.png")
```


```{r}
plot_mean_chr_expr_lineplot(infercnv_obj, sep_obs_types = TRUE)
```






