---
title: "Visualizing Large-scale Copy Number Variation in Single-Cell RNA-Seq Expression Data"
author: 
- name: Timothy Tickle
  affiliation: &kco Klarman Cell Observatory, Broad Institute of MIT and Harvard, Cambridge, MA, USA
- name: Itay Tirosh
  affiliation: 
    - *kco
    - Weizmann Institute of Science, Rehovot, Israel
- name: Brian Haas
  affiliation: *kco
output:
  BiocStyle::pdf_document: default
package: inferCNV
abstract: >
  InferCNV is used to explore tumor single cell RNA-Seq data to identify evidence for large-scale chromosomal copy number variations, such as gains or deletions of entire chromosomes or large segments of chromosomes. This is done by exploring expression intensity of genes across positions of the genome in comparison to the average or a set of reference 'normal' cells. A heatmap is generated illustrating the relative expression intensities across each chromosome, and it becomes readily apparent as to which regions of the genome are over-abundant or less-abundant as compared to normal cells (or the average, if reference normal cells are not provided).
vignette: >
  %\VignetteIndexEntry{Visualizing Large-scale Copy Number Variation in Single-Cell RNA-Seq Expression Data}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---


# Prerequisites

_inferCNV_ uses the _R_ packages `r CRANpkg("ape")`, `r CRANpkg("RColorBrewer")` and `r CRANpkg("GMD")`

<!--
[https://cran.r-project.org/web/packages/ape/index.html]
[https://cran.r-project.org/web/packages/RColorBrewer/index.html]
[https://cran.r-project.org/src/contrib/Archive/GMD/]
-->



# Run the example

```{r}
library(infercnv)
infercnv(x="../example/oligodendroglioma.tp100k.expr.matrix",
        gene_order="../example/gencode_v19_gene_pos.txt", 
        annotations="../example/__sampled_cells_bis.annotations.dat", 
        cutoff=1, 
        noise_filter=0.2, 
        output_dir="../devel_debug_bioc", 
        name_ref_groups="Microglia/Macrophage,Oligodendrocytes (non-malignant)", 
        hclust_method="ward.D",
        log_level="WARNING")


```


